jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutorWithPool.java

```diff
@@ -0,0 +1,184 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.thread;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.Exchanger;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.util.Pool;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+@ManagedObject("An alternate pool for reserved threads")
+public class ReservedThreadExecutorWithPool extends AbstractLifeCycle implements TryExecutor
+{
+    private static final Logger LOG = Log.getLogger(ReservedThreadExecutorWithPool.class);
+
+    private final Pool<ReservedThread> stack;
```

Comment:
Using the `Pool` for the stack is a big change in semantics.  Currently one a reserved thread is given a task, it is no longer counted as a reserved thread and does not hold any of the capacity, thus it does not prevent other threads becoming reserved.   Using a pool will put an absolute limit on the numer of tasks that can be reserved plus can be executed by something that used to be reserved.

I am also concerned that the underlying COW list might not be the right data structure if the reserved thread collection is moderately volatile.